From 30fee6fccebdcfc8b61745fffc3472fb1ea1c3f0 Mon Sep 17 00:00:00 2001
From: "Gadiraju, Ashokkumar" <agadiraju@irobot.com>
Date: Thu, 27 Jun 2019 09:32:04 -0400
Subject: [PATCH 2/2] sem: unmap semaphore resources when refcount is zero in
 close

unmap semaphore resources in sem_close only if reference count is zero

unmap with ref count non zero leads to user space application crashes
for counting semaphores
---
 src/thread/sem_open.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/src/thread/sem_open.c b/src/thread/sem_open.c
index e731227e..058558f6 100644
--- a/src/thread/sem_open.c
+++ b/src/thread/sem_open.c
@@ -169,11 +169,14 @@ int sem_close(sem_t *sem)
 		return -1;
 	}
 	if (!--semtab[i].refcnt) {
+		errno = -munmap(sem, sizeof *sem);
+		if (errno) {
+			UNLOCK(lock);
+			return -1;
+		}
 		semtab[i].sem = 0;
 		semtab[i].ino = 0;
 	}
 	UNLOCK(lock);
-	errno = -munmap(sem, sizeof *sem);
-	if (errno) return -1;
 	return 0;
 }
-- 
2.22.0.214.g8dca754b1e

